### DISCUSSION TITLE
Handling Sign Up with Existing Email in Supabase

### DISCUSSION THREAD

**Discussion by vladimirzb** [supabase/supabase#16770] [Questions]
I'm working on a sign-up feature using Supabase and I want to provide feedback to the user if they try to sign up with an email that already exists in the system. However, I'm facing a challenge in detecting this specific case.

When I attempt to sign up with an email that already exists, I receive the following response:

```json
{
  "data": {
    "session": null,
    "user": {
      "id": "b84ad9fb-9f81-42c4-9c5b-3d08b5e04156",
      "aud": "authenticated",
      "role": "authenticated",
      "email": "example@email.com",
      "phone": "",
      ...
    }
  },
  "error": null
}
```

I understand that Supabase returns this response to avoid leaking information about existing email addresses, but I'm unsure how to handle this case in my application.

1. Is there a recommended way to detect that the email is already in use and provide appropriate feedback to the user?
2. Should I use a condition like `if (error == null)` to handle this case, or is there a more specific approach I should take?

I appreciate any guidance or best practices for handling this scenario in a secure and user-friendly way. Thank you!

**Comment by GaryAustin1** [comment]
You could use an rpc call to a security definer function and send an email address. The check if the email exists in auth.users with SQL and return true or false. 

Searching discussions also finds this long thread.  Towards the end some other ideas.  https://github.com/orgs/supabase/discussions/1282

**Reply by vladimirzb** [reply]
Hello,

Thank you for directing me to the discussion thread. It was very helpful. I decided to take an approach similar to what PrabhdeepSingh suggested in the thread. Here's how I implemented it in my updated code:

```javascript
const handleSignUp = async () => {
  const response = await supabase.auth.signUp({
    email,
    password,
    options: {
      emailRedirectTo: `${location.origin}/auth/callback`,
    },
  });

  console.log("Sign-up response:", JSON.stringify(response, null, 2)); // Log the entire sign-up response

  if (response.data && response.data.user) {
    // Log the identities array
    console.log("Identities array:", JSON.stringify(response.data.user.identities, null, 2));

    // Check if the user got created
    if (response.data.user.identities && response.data.user.identities?.length > 0) {
      // Success, the user was created
      console.log("Sign-up successful!");
    } else {
      // Failed, the email address is taken
      console.log("Email address is already taken.");

      // Resend the confirmation email
      const resendResponse = await supabase.auth.resend({ email, type: "signup" });

      console.log("Resend response:", JSON.stringify(resendResponse, null, 2)); // Log the resend response

      if (resendResponse.error) {
        console.error("An error occurred while resending the email:", resendResponse.error.message);
      } else {
        console.log("Confirmation email resent successfully!");
      }
    }
  } else {
    // Handle other errors
    console.error("An error occurred during sign-up:", response.error?.message);
  }

  router.refresh();
};
```

Here are the logs I received (email and ID are censored for privacy):

```plaintext
Sign-up response: {
  "data": {
    "user": {
      "id": "[CENSORED]",
      "aud": "authenticated",
      "role": "authenticated",
      "email": "[CENSORED]",
      ...
      "identities": [],
      ...
    },
    "session": null
  },
  "error": null
}
Identities array: []
Email address is already taken.
Resend response: {
  "data": {
    "user": null,
    "session": null
  },
  "error": null
}
Confirmation email resent successfully!
```

I noticed that the confirmation email is not being resent even though the console logs indicate "Confirmation email resent successfully!". Do you have any insights into why this might be happening?

Thank you for your time and assistance!

**Reply by igorlanko** [reply]
Hey @vladimirzb, I'm having similar issues with `resend()`. No errors on the client or Supabase API, but Inbucket is not seeing any new emails resent.

What was the cause in your case?
